From 59e241d64639a6d56ebcd1aa2a9d60678a849373 Mon Sep 17 00:00:00 2001
From: Drew Moseley <drew.moseley@northern.tech>
Date: Sun, 10 Dec 2017 14:00:54 -0500
Subject: [PATCH] Enable U-Boot on QEMU to use Mender boot code.

Also make sure the U-Boot environment is on SD card.

Signed-off-by: Drew Moseley <drew.moseley@northern.tech>
---
 configs/vexpress_ca9x4_defconfig  |  3 ++-
 include/configs/vexpress_common.h | 24 ++++++++++++++++++++----
 2 files changed, 22 insertions(+), 5 deletions(-)

diff --git a/configs/vexpress_ca9x4_defconfig b/configs/vexpress_ca9x4_defconfig
index 4cbc962..d7f19b7 100644
--- a/configs/vexpress_ca9x4_defconfig
+++ b/configs/vexpress_ca9x4_defconfig
@@ -16,7 +16,7 @@ CONFIG_CMD_MMC=y
 # CONFIG_CMD_SETEXPR is not set
 # CONFIG_CMD_NFS is not set
 # CONFIG_CMD_MISC is not set
-CONFIG_ENV_IS_IN_FLASH=y
+# CONFIG_ENV_IS_IN_FLASH is not set
 CONFIG_MTD_NOR_FLASH=y
 CONFIG_NETDEVICES=y
 CONFIG_SMC911X=y
@@ -24,3 +24,4 @@ CONFIG_SMC911X_BASE=0x4e000000
 CONFIG_SMC911X_32_BIT=y
 CONFIG_BAUDRATE=38400
 CONFIG_OF_LIBFDT=y
+CONFIG_ENV_IS_IN_MMC=y
\ No newline at end of file
diff --git a/include/configs/vexpress_common.h b/include/configs/vexpress_common.h
index ade7ba4..fe81dac 100644
--- a/include/configs/vexpress_common.h
+++ b/include/configs/vexpress_common.h
@@ -158,6 +158,9 @@
 #define CONFIG_SYS_LOAD_ADDR		(V2M_BASE + 0x8000)
 #define LINUX_BOOT_PARAM_ADDR		(V2M_BASE + 0x2000)
 
+#define CONFIG_BOOTCOUNT_LIMIT
+#define CONFIG_BOOTCOUNT_ENV
+
 /* Physical Memory Map */
 #define CONFIG_NR_DRAM_BANKS		2
 #define PHYS_SDRAM_1			(V2M_BASE)	/* SDRAM Bank #1 */
@@ -188,6 +191,22 @@
         func(DHCP, dhcp, na)
 #include <config_distro_bootcmd.h>
 
+#undef CONFIG_BOOTCOMMAND
+#define CONFIG_BOOTCOMMAND \
+    "run mender_setup; " \
+    "run qemu_boot"
+
+#define MENDER_QEMU_ENV_SETTINGS                                        \
+    "qemu_kernel_addr=0x60100000\0"                                     \
+                                                                        \
+    "qemu_dtb_addr=0x60000000\0"                                        \
+                                                                        \
+    "qemu_boot=load ${mender_uboot_root} ${qemu_kernel_addr} /boot/zImage; " \
+    "load ${mender_uboot_root} ${qemu_dtb_addr} /boot/vexpress-v2p-ca9.dtb; " \
+    "setenv bootargs root=${mender_kernel_root} console=tty0 console=ttyAMA0,115200n8; " \
+    "bootz ${qemu_kernel_addr} - ${qemu_dtb_addr} ; "                   \
+    "reset\0"
+
 #ifdef CONFIG_VEXPRESS_ORIGINAL_MEMORY_MAP
 #define CONFIG_PLATFORM_ENV_SETTINGS \
 		"loadaddr=0x80008000\0" \
@@ -211,6 +230,7 @@
 #endif
 #define CONFIG_EXTRA_ENV_SETTINGS \
 		CONFIG_PLATFORM_ENV_SETTINGS \
+		MENDER_QEMU_ENV_SETTINGS \
                 BOOTENV \
 		"console=ttyAMA0,38400n8\0" \
 		"dram=1024M\0" \
@@ -256,10 +276,6 @@
 #define CONFIG_ENV_OVERWRITE		1
 
 /* Store environment at top of flash */
-#define CONFIG_ENV_OFFSET		(PHYS_FLASH_SIZE - \
-					(2 * CONFIG_ENV_SECT_SIZE))
-#define CONFIG_ENV_ADDR			(CONFIG_SYS_FLASH_BASE1 + \
-					 CONFIG_ENV_OFFSET)
 #define CONFIG_SYS_FLASH_PROTECTION	/* The devices have real protection */
 #define CONFIG_SYS_FLASH_EMPTY_INFO	/* flinfo indicates empty blocks */
 #define CONFIG_SYS_FLASH_BANKS_LIST	{ CONFIG_SYS_FLASH_BASE0, \
-- 
2.7.4

